### PR TITLE
Port semantic tests to the new test tool

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Ports the semantic tests written in DAML to the new test tool.

Drops all the machinery needed to evaluate scenarios from within the test tool (>1K lines of complicated code dropped, me likes this very much).

Note that there is some overlapping between the semantic tests and the existing tests as ported from the integration tests (e.g. most likely contract keys and divulgence test are already covered): still, I preferred porting the tests on a 1:1 basis to not break the validity of existing test runs.

Some work will be needed to streamline and organize the tests in the future.

**DO NOT MERGE (my finger slipped when I was about to create a draft PR), this builds on top of unmerged work.**

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala [hunk 1]
```diff
@@ -3,52 +3,340 @@
...
+      "SemanticAcceptOnBehalf",
+      "It should not be possible to exercise a choice without the consent of the controller") {
+      ledger =>
+        for {
+          Vector(bank, houseOwner, painter) <- ledger.allocateParties(3)
+          iou <- ledger.create(bank, Iou(bank, houseOwner, onePound))
+          offer <- ledger.create(painter, PaintOffer(painter, houseOwner, bank, onePound))
+          failure <- ledger.exercise(painter, offer.exercisePaintOffer_Accept(_, iou)).failed
+        } yield {
+          assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "couldn't find contract")
```
Comment (meiersi-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: The error code looks wrong here. We'd want an authorization failure and not a privacy failure. The problem is that this `Iou` is not visible to the painter.

An easy way out is to add the option to `Reject` the offer for the houseOwner and try to trigger that rejection as the painter. In that case no Iou needs to be created.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala [hunk 2]
```diff
@@ -3,52 +3,340 @@
...
+        _ <- fetchIou(ledger, painter, iou).failed
+
+        offer <- ledger.create(painter, PaintOffer(painter, houseOwner, bank, onePound))
+
+        // The house owner and the painter can see the offer but the bank can't
+        _ <- fetchPaintOffer(ledger, houseOwner, offer)
+        _ <- fetchPaintOffer(ledger, painter, offer)
+        _ <- fetchPaintOffer(ledger, bank, offer).failed
+
+        // Note that the new IOU was divulged to the house owner
```
Comment (meiersi-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: @stefanobaghino-da : it does not seem to me that we are exploiting this note, are we?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala [hunk 3]
```diff
@@ -3,52 +3,340 @@
...
+   */
+
+  private[this] val divulgence =
+    LedgerTest("SemanticDivulgence", "Respect divulgence rules") { ledger =>
+      for {
+        Vector(issuer, owner, delegate) <- ledger.allocateParties(3)
+        token <- ledger.create(issuer, Token(issuer, owner, 1))
+        delegation <- ledger.create(owner, Delegation(owner, delegate))
+
+        // The owner tries to delegate with a non-consuming choice, which actually doesn't
```
Comment (meiersi-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: typo `s/doesn't/doesn't work`

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala [hunk 4]
```diff
@@ -3,52 +3,340 @@
...
+   * Contract keys
+   */
+
+  private[this] val contractKeys =
+    LedgerTest("SemanticContractKeys", "Perform correctly operations based on contract keys") {
+      ledger =>
+        for {
+          Vector(bank, accountHolder) <- ledger.allocateParties(2)
+          accountTemplate = Account(bank, accountHolder, DamlTuple2("CH", 123))
+          accountKey = DamlTuple2(bank, "CH123") // key (bank, accountNumber._1 <> show (this.accountNumber._2)) : (Party, Text)
```
Comment (meiersi-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: @stefanobaghino-da : what does the comment here mean?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: Addressed by https://github.com/digital-asset/daml/pull/2718/commits/c0e8912d2ea892eaa3ff0f3e22c2abdc39150d6f

The cause was that the painter is unable to see the `iou`. I worked around the problem by giving the `iou` to the painter.

This is a very good catch.

giphy

This actually highlights a possible problem in how we do the validation, as @gerolf-da just told me offline.

I'd say we can (for now) limit ourselves to make sure the test passes but we can easily add another test that checks that the validation is performed in the correct order.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: Also `s/delegate/divulge`. Fixing, thanks.

(FYI, point 6 [here](https://help.github.com/en/articles/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) can come in handy for this kind of comments :wink:)

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala (hunk 4 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: It's the way in which the key is computed on DAML. I left it there for reference. Unfortunately I don't think there is a way to do it organically without replicating the logic to get the key. Not that readable though, thanks for the comment, I'll make it explicit.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: Address by https://github.com/digital-asset/daml/pull/2718/commits/d3e4ae9db78fec56118cae9c519a3aea8f728c7c

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala (hunk 4 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: Addressed by https://github.com/digital-asset/daml/pull/2718/commits/7a623c3d5de46231da70f08dc86df5621b56bdbc

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: Addressed by https://github.com/digital-asset/daml/pull/2718/commits/d3e4ae9db78fec56118cae9c519a3aea8f728c7c

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: Addressed by https://github.com/digital-asset/daml/pull/2718/commits/a97a9e24a4c8fb23a5ba007bb7204e5987356b1c

Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala]: Addressed by https://github.com/digital-asset/daml/pull/2718/commits/a97a9e24a4c8fb23a5ba007bb7204e5987356b1c

**Comment by stefanobaghino-da** [comment]
Squashed and rebased on `master` after #2713 got merged, will merge as soon as I receive a positive feedback from CI.

**Pull request merged**